File: networksML.py
import torch
import numpy as np
import matplotlib.pyplot as plt
import typing

# from numba import jit
from torch import nn
# from scipy.optimize import minimize
from multiprocessing import Pool, cpu_count

rng = np.random.default_rng()


def get_dict(s: list) -> dict:
    out = {j: [] for j in set(x[0] for x in s)}
    for x in s:
        out[x[0]].append((x[1], x[2]))
    return out


class Network:

    def __init__(self, graph: np.ndarray, value_mean: float = 0.0, value_sd: float = 1.0, signal_life: int = 5, signal_error_sd: float = 0.1):
        self.graph = graph
        assert(graph.shape[0] == graph.shape[1])
        self.graph_size = graph.shape[0]
        self.values = rng.normal(loc=value_mean, scale=value_sd, size=self.graph_size)
        self.signal_life = signal_life
        self.signal_error_sd = signal_error_sd
        self._set_signals()
    
    def _set_signals(self):
        signals = [[] for _ in range(self.graph_size)]
        # format of signal is (origin, direct source, value)
        to_send = [[(i, i, self.values[i])] for i in range(self.graph_size)]
        # keep track of where signals are coming from
        # only first signal from a given source about a given member will be kept
        unique_sources = [[(i, i)] for i in range(self.graph_size)]
        for _ in range(self.signal_life):
            received = [[] for _ in range(self.graph_size)]
            for i, to_send_ in enumerate(to_send):
                for j in range(self.graph_size):
                    if i != j and self.graph[i, j]:
                        warped_signals = [
                            (s[0], i, s[2] + rng.normal(scale=self.signal_error_sd))
                            for s in to_send_ if s[0] != j and (s[0], i) not in unique_sources[j]
                        ]
                        received[j] += warped_signals
                        unique_sources[j] += [(s[0], s[1]) for s in warped_signals]
            for sig, sent in zip(signals, to_send):
                sig += sent
            to_send = received
        # format as dict
        self.signals = [get_dict(s) for s in signals]


def gen_random_graph(size: int, p: float = 0.5) -> np.ndarray:
    """Creates a symmetric matrix of size (size x size) where proba of [off-diagonal element == True] is p"""
    random_mat = rng.random((size, size))
    return (random_mat + random_mat.T < 2 * p).astype(bool) * ~np.eye(size, dtype=bool)


# now set up neural nets and training algorithm

def init_weights(m):
    if isinstance(m, nn.Linear):
        torch.nn.init.xavier_uniform_(m.weight)
        m.bias.data.fill_(0.00)


class MLModel(nn.Module):

    def __init__(self, graph_size: int, max_signals: int, n_hidden: int = 4, hidden_size: int = 50):
        super().__init__()
        
        self.graph_size = graph_size
        self.max_signals = max_signals
        
        # input features are:
        # graph_size**2 features to describe the graph
        # graph_size to indicate guesser id
        # for each graph member * for each of max_signals:
        # graph_size features to indicate signal sender id
        # 1 for signal value
        self.n_other_features = graph_size * self.max_signals * (graph_size + 1)
        self.n_features = self.graph_size**2 + self.graph_size + self.n_other_features
        
        self.first = nn.Linear(self.n_features, hidden_size)
        self.hidden = nn.ModuleList([
            nn.Linear(hidden_size, hidden_size) for _ in range(n_hidden)
        ])
        self.last = nn.Linear(hidden_size, graph_size)

        self.apply(init_weights)
    
    def get_features(self, net: Network) -> torch.Tensor:
        assert(net.graph_size == self.graph_size)
        graph_features = torch.tile(
            torch.from_numpy(net.graph.flatten()), (self.graph_size,1)
        )
        id_features = torch.eye(self.graph_size)
        other_features = torch.zeros((self.graph_size, self.n_other_features))
        # iterate through signal receivers
        for i, signals in enumerate(net.signals):
            # iterate through signal subjects
            for j, s in signals.items():
                start_idx = self.max_signals * (self.graph_size + 1) * j
                # iterate_through signal senders
                for idx, (sender, value) in enumerate(s):
                    if idx >= self.max_signals:
                        break
                    offset = start_idx + (self.graph_size + 1) * idx
                    other_features[i, offset + sender] = 1.0
                    other_features[i, offset + self.graph_size] = value
        return torch.concat((graph_features, id_features, other_features), axis=-1).float()
    
    def forward(self, x: torch.Tensor) -> torch.Tensor:
        x = torch.relu(self.first(x))
        for h in self.hidden:
            x = x + torch.relu(h(x))
        return self.last(x)

    def train(
        self,
        connection_proba: float = 0.5,
        value_mean: float = 0.0, value_sd: float = 1.0, signal_life: int = 5, signal_error_sd: float = 0.1,
        epochs: int = 100, lr: float = 1e-3,
        epoch_size: int = 20, threads: int = min(20, cpu_count())
    ) -> list:
        optimizer = torch.optim.Adam(self.parameters(), lr = lr)
        loss_fn = nn.MSELoss()
        loss_history = []
        for i in range(epochs):
            feature_target_list = []
            while len(feature_target_list) < epoch_size:
                pool_size = min(threads, epoch_size - len(feature_target_list))
                with Pool(pool_size) as pool:
                    feature_target_list += pool.map(
                        _get_features,
                        [(self, connection_proba, value_mean, value_sd, signal_life, signal_error_sd)] * pool_size
                    )
                    
            features = torch.concat(tuple(f[0] for f in feature_target_list))
            targets = torch.concat(tuple(f[1] for f in feature_target_list))
            optimizer.zero_grad()
            loss = loss_fn(self.forward(features), targets)
            loss_history.append(loss.item())
            loss.backward()
            optimizer.step()
            if ((i+1) % (epochs // 10) == 0):
                print(f'Epoch {i+1} of {epochs}: Loss = {loss:.3e}')
        return loss_history


def get_features(
    model: MLModel,
    connection_proba: float,
    value_mean: float, value_sd: float, signal_life: int, signal_error_sd: float
) -> typing.Tuple[torch.Tensor, torch.Tensor]:
    graph = gen_random_graph(model.graph_size, connection_proba)
    net = Network(graph, value_mean, value_sd, signal_life, signal_error_sd)
    features = model.get_features(net)
    targets = torch.tile(torch.from_numpy(net.values), (model.graph_size, 1)).float()
    return features, targets

def _get_features(
    args: typing.Tuple[MLModel, float, float, float, int, float]
) -> typing.Tuple[torch.Tensor, torch.Tensor]:
    return get_features(*args)


def main():
    """Currently just a test to run to make sure things are working well"""
    model = MLModel(12, 2)

    loss_history = model.train(connection_proba=0.2, lr=1e-2, epochs=100)

    plt.plot(loss_history, linestyle='', marker='.', alpha=0.6)
    plt.show()

    graph = gen_random_graph(model.graph_size, 0.5)
    net = Network(graph)

    with torch.no_grad():
        features = model.get_features(net)
        preds = model.forward(features)

    print(graph)
    print(net.values)
    print(preds.numpy())


if __name__ == '__main__':
    main()

File: networks_ml.cpp
#include <torch/torch.h>
#include <cstddef>
#include <memory>
#include <vector>
#include <string>
#include <thread>
#include <random>
#include <assert.h>

#include <iostream>


std::default_random_engine rng;


template <typename T>
class Matrix {
public:
    Matrix(std::vector<T> values, std::size_t size) : values_(values), size_(size) {
        assert(values_.size() == size * size);
    }

    std::size_t size() const {
        return size_;
    }

    T get(std::size_t i, std::size_t j) const {
        assert(i < size_ && j < size_);
        return values_[i * size_ + j];
    }

    void set(std::size_t i, std::size_t j, T new_val) {
        assert(i < size_ && j < size_);
        values_[i * size_ + j] = new_val;
    }

    std::vector<T> values() const {
        return values;
    }

    void print() const {
        for (std::size_t i = 0; i < size_; i++) {
            for (std::size_t j = 0; j < size_; j++) {
                std::cout << (get(i, j)) ? '1' : '0';
                if (j == size_ - 1) {
                    std::cout << '\n';
                }
                else {
                    std::cout << ' ';
                }
            }
        }
    }

private:
    std::vector<T> values_;
    std::size_t size_;
};

Matrix<bool> gen_random_graph(std::size_t size, double p) {
    assert(0.0 <= p <= 1.0);
    std::vector<bool> values;
    values.reserve(size * size);
    std::uniform_real_distribution<double> dist(0, 1);
    for (std::size_t i = 0; i < size * (size - 1) / 2; i++) {
        values.push_back(dist(rng));
    }
    return Matrix<bool>(values, size);
}

std::vector<double> sample_normal(double mean, double sd, std::size_t size) {
    std::vector<double> out;
    std::normal_distribution<double> dist(mean, sd);
    out.reserve(size);
    for (std::size_t i = 0; i < size; i++) {
        out.push_back(dist(rng));
    }
    return out;
}


struct Signal {
    Signal(
        std::size_t sender, std::size_t source, double value
    ) : sender(sender), source(source), value(value) {}
    std::size_t sender;
    std::size_t source;
    double value;
};


class NetMember {
public:
    NetMember(
        std::size_t graph_size, std::size_t my_idx, double my_value
    ) : graph_size(graph_size), my_idx(my_idx), to_share({Signal(my_idx, my_idx, my_value)}) {
        signals.resize(graph_size);
    }

    void receive_signal(Signal signal) {
        if (is_new_source_and_origin(signal)) {
            signals[signal.source].push_back(signal);
            recently_received.push_back(signal);
        }
    }

    void flush_recent_signals() {
        to_share = recently_received;
        recently_received = {};
    }

    std::size_t get_my_idx() const {
        return my_idx;
    }

    const std::vector<Signal>& get_signals_to_share() const {
        return to_share;
    }

    void print() const {
        std::cout << "Signals received by member " << my_idx << ":\n";
        for (const auto& signal_set : signals) {
            for (const Signal& signal : signal_set) {
                std::cout << "Source: " << signal.source
                    << ", Sender: " << signal.sender
                    << ", Value: " << signal.value << '\n';
            }
        }
        std::cout << '\n';
    }

private:
    bool is_new_source_and_origin(const Signal& signal) const {
        if (signal.source == my_idx) {
            // This signal is about me, so I don't need it
            return false;
        }
        for (const Signal& my_signal : signals[signal.source]) {
            if (my_signal.sender == signal.sender) {
                // I already have a signal about this source from this sender
                return false;
            }
        }
        return true;
    }

    std::size_t graph_size;
    std::size_t my_idx;

    std::vector<std::vector<Signal>> signals;
    std::vector<Signal> to_share;
    std::vector<Signal> recently_received;
};


class Network {
public:
    Network(
        Matrix<bool> graph,
        double value_mean,
        double value_sd,
        unsigned int signal_life,
        double signal_error_sd
    ) : graph_(graph),
        signal_life(signal_life),
        signal_error_sd(signal_error_sd),
        values_(sample_normal(value_mean, value_sd, graph.size())) 
    {
        set_signals();
    }

    Matrix<int> get_distances() const {
        std::size_t size = graph_.size();
        // organize index of neighbors & create a distances matrix based on graph
        std::vector<int> values;
        values.reserve(size * size);
        std::vector<std::vector<std::size_t>> neighbors(size);
        for (std::size_t i = 0; i < size; i++) {
            for (std::size_t j = 0; j < size; j++) {
                bool graph_val = graph_.get(i, j);
                int distance = (i == j) ? 0 : ((graph_val) ? 1 : -1);
                values.push_back(distance);
                if (graph_.get(i, j)) {
                    neighbors[i].push_back(j);
                }
            }
        }
        Matrix<int> distances(values, size);

        unsigned int n = 1;
        bool keep_going = true;
        while (keep_going) {
            keep_going = false;
            for (std::size_t i = 0; i < size; i++) {
                for (std::size_t j = 0; j < size; j++) {
                    if (i == j || distances.get(i, j) != n) {
                        continue;
                    }
                    keep_going = true;
                    for (auto k : neighbors[j]) {
                        if (i != k && distances.get(i, k) == -1) {
                            distances.set(i, k, n + 1);
                        }
                    }
                }
            }
            n++;
        }
        return distances;
    }

    void print_signals() {
        for (const NetMember& member : net_members_) {
            member.print();
        }
    }

    const Matrix<bool>& graph() const {
        return graph_;
    }

    const std::vector<double>& values() const {
        return values_;
    }

    const std::vector<NetMember>& net_members() const {
        return net_members_;
    }

private:
    void set_signals() {
        // get things set up
        std::size_t size = graph_.size();
        std::vector<NetMember> members;
        members.reserve(size);
        for (std::size_t i = 0; i < size; i++) {
            members.push_back(NetMember(size, i, values_[i]));
        }
        std::normal_distribution<double> noise_dist(0.0, signal_error_sd);

        // now fill in signals
        // loop through time periods
        for (unsigned int t = 0; t < signal_life; t++) {
            // loop through signal senders
            for (std::size_t i = 0; i < size; i++) {
                const auto& signals_to_share = members[i].get_signals_to_share();
                // loop through signal receivers
                for (std::size_t j = 0; j < size; j++) {
                    if (i == j || !graph_.get(i, j)) {
                        continue;
                    }
                    for (const Signal& s : signals_to_share) {
                        members[j].receive_signal(Signal(i, s.source, s.value + noise_dist(rng)));
                    }
                }
                members[i].flush_recent_signals();
            }
        }

        net_members_ = members;
    }

    Matrix<bool> graph_;
    unsigned int signal_life;
    double signal_error_sd;
    std::vector<double> values_;
    std::vector<NetMember> net_members_;
};



void xavier_init(torch::nn::Module& module) {
	torch::NoGradGuard noGrad;
	if (auto* linear = module.as<torch::nn::Linear>()) {
		torch::nn::init::xavier_normal_(linear->weight);
		torch::nn::init::constant_(linear->bias, 0.01);
	}
}


struct MLModel : torch::nn::Module {
    MLModel(
        // graph params
        std::size_t graph_size,
        double connection_proba = 0.5,
        double value_mean = 0.0,
        double value_sd = 1.0,
        std::size_t signal_life = 5,
        double signal_error_sd = 0.1,
        // model params
        std::size_t max_signals = 4,
        std::size_t n_hidden = 4,
        std::size_t hidden_size = 50
    ) : graph_size(graph_size),
        max_signals(max_signals),
        connection_proba(connection_proba),
        value_mean(value_mean),
        value_sd(value_sd),
        signal_life(signal_life),
        signal_error_sd(signal_error_sd)
    {
        // input features are:
        // graph_size**2 features to describe the graph
        // graph_size to indicate guesser id
        // for each graph member * for each of max_signals:
        // graph_size features to indicate signal sender id
        // 1 for signal value
        n_other_features = graph_size + max_signals + (graph_size + 1);
        n_features = graph_size * (graph_size + 1) + n_other_features;

        first = register_module(
            "first",
            torch::nn::Linear(n_features, hidden_size)
        );
        for (std::size_t i = 0; i < n_hidden; i++) {
            hidden.push_back(
                register_module(
                    "hidden_" + std::to_string(i),
                    torch::nn::Linear(hidden_size, hidden_size)
                )
            );
        }
        last = register_module(
            "last",
            torch::nn::Linear(hidden_size, graph_size)
        );

        this->apply(xavier_init);
    }

    torch::Tensor get_features(const Network& net) {
        assert(net.graph().size() == graph_size);
        auto graph_features = torch::tile(
            torch::from_blob(
                net.graph().values(), torch::dtype(torch::kFloat32)
            ),
            {graph_size, n_other_features}
        );
        auto id_features = torch::eye(graph_size);
        auto other_features = torch::zeros({graph_size, n_other_features});
        // iterate through network members
        auto net_members = net.net_members();
        for (std::size_t i = 0; i < net_members.size(); i++) {
            // iterate through signal sources
            auto signals_set = net_members[i].signals;
            for (std::size_t j = 0; j < signals_set.size(); j++) {
                const auto& signals = signals_set[j];
                std::size_t start_idx = max_signals * (graph_size + 1) * j;
                // iterate through signal senders
                for (
                    std::size_t k = 0;
                    k < signals.size() && k < max_signals;
                    k++
                ) {
                    const Signal& signal = signals[k];
                    std::size_t offset = start_idx + (graph_size + 1) * k;
                    other_features.index({i, offset + signal.sender}) = 1.0;
                    other_features.index({i, offset + graph_size}) = signal.value
                }
            }
        }
        return torch::concat({graph_features, id_features, other_features}, 1);
    }

    torch::Tensor forward(torch::Tensor x) {
        x = torch::relu(first->forward(x));
        for (auto& h : hidden) {
            x = x + torch::relu(h->forward(x));
        }
        return last->forward(x);
    }

    std::vector<double> train(
        std::size_t epochs = 100,
        double lr = 0.001
        // std::size_t epoch_size = 20,
        // std::size_t threads = 20
    ) {
        auto optimizer = torch::optim::Adam(parameters(), lr);
        loss_fn = torch::nn::MSELoss();
        std::vector<double> loss_history;
        loss_history.reserve(epochs);
        for (std::size_t i = 0; i < epochs; i++) {
            auto graph = gen_random_graph(graph_size, connection_proba);
            Network net(graph, value_mean, value_sd, signal_life, signal_error_sd);
            auto features = get_features(net);
            auto targets = torch::tile(
                torch::from_blob(net.values),
                {graph_size, 1}
            );
            optimizer.zero_grad();
            auto loss = loss_fn(forward(features), targets);
            loss_history.push_back(loss.item<double>());
            loss.backward();
            optimizer.step();
            if ((i + 1) % epochs == 0 || i + 1 == epochs) {
                std::cout << "Epoch " << i + 1 << " of " << epochs << ": Loss = " << loss.item<double>() << '\n';
            }
        }
        return loss_history;
    }

    std::size_t graph_size;
    std::size_t max_signals;
    std::size_t signal_life;
    std::size_t signal_error_sd;
    std::size_t n_other_features;
    std::size_t n_features;
    double connection_proba;
    double value_mean;
    double value_sd;

    torch::nn::Linear first;
    std::vector<torch::nn::Linear> hidden;
    torch::nn::Linear last;

};


int main() {
    // Matrix<bool> graph(
    //     {
    //         false, true, false,
    //         true, false, true,
    //         false, true, false
    //     },
    //     3
    // );
    auto graph = gen_random_graph(3, 0.5);
    graph.print();

    Network net(graph, 0.0, 1.0, 4, 0.1);
    auto distances = net.get_distances();
    distances.print();

    net.print_signals();

    MLModel model(3);
    model.train();
}
